Feature: Extra Networks Tree View

Sj-Si opened the pull request:
## Latest Update

Most recent major update to this PR: https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/14588#issuecomment-1894425741

## Description

This PR adds an option to the `Extra Networks` section to switch from the "Card View" to a "Tree View". This allows users to easily navigate more complex directory structures in order to find networks.

Along with this, a new button has been added to both the existing Cards in the extra networks tabs and to the entries in this new tree view. This button (shown in the screenshot) can be clicked to copy the file path of the network to the clipboard.

Finally, if an Extra Networks tab's corresponding root directory (in `stable-diffusion-webui/models`) is empty, that directory will remain expanded and display the message `DIRECTORY IS EMPTY` (see the screenshots below).

I made these changes after getting sick of the existing Card View format for these networks. They become too cumbersome when you have many networks in a more complex directory structure.

## Screenshots/videos:

Shown below is the new entry in the `Extra Networks` section of the settings page. This setting is off by default so the user will have to actually go and change this setting to see any of the changes in this PR.

![pr_screenshot_1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/f743b4a6-e2a1-4c59-aa06-6aa70f332cef)

The following screenshot shows what the Extra Networks tabs look like with this new change enabled in the settings.

![pr_screenshot_0](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/63df2276-a9ea-42f8-8574-e4309ccabca5)

The folders that appear in these tabs are collapsed by default and can be clicked to expand.

Additionally, this is the new `Copy Path` button that was added to the existing network cards.
![pr_screenshot_2](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/2c15fe16-a272-4bc8-ad60-3e03defc88cb)

## Checklist:

- [x] I have read [contributing wiki page](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing)
- [x] I have performed a self-review of my own code
- [x] My code follows the [style guidelines](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing#code-style)
- [x] My code passes [tests](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Tests)
- [x] Tested on Ubuntu (via WSL) and Windows10

## Additional Notes:

The `Search` and `Sort` options are not implemented in the tree view. If anyone wants to tackle that beast then be my guest.

This is pretty low priority and I really just made it for my own use anyway so no rush.

Sj-Si committed:
Add tree-view display for extra networks.

Sj-Si committed:
fix indentation

Sj-Si committed:
fix linting

Sj-Si commented:
Should I be resolving conflicts with the `dev` branch? Or should I only be resolving conflicts against the `master` branch?

Sj-Si commented:
> Only dev. PRs are to be merged with the dev. People generally don't resolve conflicts, I do that, but if you do, that would help.

Understood. I'll resolve the conflicts.

It could be a load off your shoulders if you require the PR creator to resolve conflicts though... :)

AUTOMATIC1111 commented:
Now, as for the feature itself, I'm not a big fan of having a second mode for extra networks page so that the user chooses to either have cards with pictures, or a folder tree. I feel a better solution is if a tree is added (if enabled in settings), into the main mode with cards: only directories are shown in the tree, and clicking on the directory sets the search filter to only show cards from this directory. Do you have any thoughts about this way to implement the tree?

Sj-Si commented:
> Now, as for the feature itself, I'm not a big fan of having a second mode for extra networks page so that the user chooses to either have cards with pictures, or a folder tree. I feel a better solution is if a tree is added (if enabled in settings), into the main mode with cards: only directories are shown in the tree, and clicking on the directory sets the search filter to only show cards from this directory. Do you have any thoughts about this way to implement the tree?

I think I get what you're asking for. Were you thinking maybe something like GitHub's `Files Changed` window where the file explorer is on the left side and the cards could be on the right?

![tmp](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/2ce5de71-6c56-402f-907b-59b7ced989c2)

Personally I think that would be the cleanest looking option. Could also chuck that search bar at the top of the file explorer like github did.

I'll work on a quick demo of this solution and get back to you.

w-e-w commented:
oh nice after all this time someone finally decide actually work on this tree view

Sj-Si committed:
begin redesign of tree module.

w-e-w commented:
I think it shouldn't even have lora at all
like you're already on the lora tab, you are in the lora dir
lora is the root

and if you are using a custom dir with `--lora-di`r there's no guarantee that what directory name is unser let alone that exist a upper level dir

Sj-Si commented:
> fourth option, none I think it shouldn't even have lora at all like you're already on the lora tab, so you are in the lora dir whitch means lora is the root
> 
> also if you are using a custom dir with `--lora-dir` there's no guarantee that what directory name is it will be in let alone that theres exist a upper level dir (yes most likely that will be a upper level directory but if someone wants they can put all the models under system root)

If someone is putting their models at system root they deserve an error haha. Good point though I didn't think about that. My concern is that the Lora tab looks inside of both the Lora and LyCORIS directories by default. What if they have a subdirectory in each of those directories with the same name? So `models/Lora/one` and `models/LyCORIS/one`. Then they would just have two folders showing up that look the same until they navigate into them; it just feels a bit odd.

w-e-w commented:
> My concern is that the Lora tab looks inside of both the Lora and LyCORIS directories by default. What if they have a subdirectory in each of those directories with the same name? So models/Lora/one and models/LyCORIS/one

irrc if you have two models of the same name at the same relative position in both Lora and LyCORIS only one will work (I forgot which one has priority)
so it is a non issue or at least this will only be an issue if extra network is updated to support this

also the LyCORIS dir is left for compatibility reasons only (you don't have to use it and I think it's better you don't use it)

if this tree view can display multiple networks from different roots with the same relative position and name then one can consider this to be a bug because extra networks doesn't support it
(I don't think we even support if you have multiple lora under the same name across subdirectories I believe only one will be used)

basically unless non-unique names and multiple root is supported by the extra network the tree should not display the networks that can't be used

Sj-Si committed:
Finish base layout. Fix bugs. Need to test for stability and clean up.

Sj-Si commented:
Okay I've updated the format a bit. Now the tree view is on a left pane alongside the cards. The filters work and when you click on directories it will use that as the search filter. Clicking on a directory will expand that directory and use that directory as a search filter for the right side of the pane. Clicking an already opened directory will just select that directory but leave it expanded. To collapse a directory, you just click that directory again if it is already selected.

Currently if your search is a directory then all networks within that directory and subdirectories will be shown. This is because the search is just checking for a substring in the search terms and relative paths are included in the search terms for each network.

I also updated the code so that switching between tabs clears the search box. Having the search filter stay the same when switching between tabs was annoying since the models don't have the same search terms.

I also decided to keep the networks in the directory tree since I personally find it much quicker to just use the tree view to select my models.

It's still unpolished because I don't know any html/css/js so this is about as good as I can get it within a reasonable time.

One thing I would have liked to see is the search textbox with an `x` in it to clear the search but I couldn't find a way to do that with a `gr.Textbox`.

Below is a quick demo video showing how clicking on directories updates the search box.

https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/251393b2-65df-4985-a73b-159c87b2e28b

The code still definitely needs to be cleaned up. I also need to resolve conflicts with `dev`.

Sj-Si committed:
Merge changes from dev

Sj-Si commented:
Hey so I'm running into an issue that I cant seem to resolve.

In `extraNetworks.js` I want to register event handlers for various elements in the page. However, when I do `gradioApp().querySelector("#my_id")` it cannot find any elements with that ID. But when I run that command in the browser console I do get elements in the result. So it seems like the `extraNetworks.js` file is being run prior to the html actually being built.

Am I correct in this assumption and is there any way to have a `.js` file that runs after the html is generated?

Sj-Si committed:
continue cleanup and redesign.

w-e-w commented:
> Am I correct in this assumption and is there any way to have a .js file that runs after the html is generated?

I think one of these callbacks might work for you

https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/cb5b335acddd126d4f6c990982816c06beb0d6ae/script.js#L27-L33

Sj-Si committed:
fix search box handling. sort maybe broken not sure. need to bug test and finish cleanup.

Sj-Si commented:
> > Am I correct in this assumption and is there any way to have a .js file that runs after the html is generated?
> 
> I think one of these callbacks might work for you
> 
> https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/cb5b335acddd126d4f6c990982816c06beb0d6ae/script.js#L27-L31

Thank you that helped. I wasn't able to use any of these predefined callbacks but I was able to figure out a solution using a similar approach to `uiAfterUpdateCallbacks`. What I found is that the content within the extra networks tabs doesn't finish loading before `uiLoadedCallbacks`. The `uiAfterUpdateCallbacks` should work but it gets called on every single update to the page (clicking tabs, buttons, etc.) so I don't really want to use that either.

```javascript
var uiAfterScriptsCallbacks = []
var uiAfterScriptsTimeout = null;
var executedAfterScripts = false;

function scheduleAfterScriptsCallbacks() {
    clearTimeout(uiAfterScriptsTimeout);
    uiAfterScriptsTimeout = setTimeout(function() {
        executeCallbacks(uiAfterScriptsCallbacks);
    }, 200);
}

document.addEventListener("DOMContentLoaded", function() {
    var mutationObserver = new MutationObserver(function(m) {
        if (!executedAfterScripts &&
            gradioApp().querySelectorAll("[id$='_extra_search']").length == 8) {
            executedAfterScripts = true;
            scheduleAfterScriptsCallbacks();
        }
    });
    mutationObserver.observe(gradioApp(), {childList: true, subtree: true});
});

uiAfterScriptsCallbacks.push(setupExtraNetworks);
```

This method waits for all of the elements matching `*_extra_search` to finish getting generated on the page before it sets up the extra networks tabs. I wish there were some better way to handle this than by checking if specific elements finish loading. Maybe adding a dummy element somewhere on the page that we can guarantee will always be loaded last could be a solution though it feels a bit hacky for my taste. I'm also not a fan of hardcoding the length check in `gradioApp().querySelectorAll("[id$='_extra_search']").length == 8)` but it works at the moment since I know that there will always be 4 of these elements generated in each of the `txt2img` and `img2img` tabs.

Sj-Si committed:
fix linting

Sj-Si committed:
fix bugs. introduce new ones.

Sj-Si committed:
Move extra network tab buttons into tree view;

Sj-Si committed:
Run linting

Sj-Si committed:
Finish cleanup.

Sj-Si committed:
fix bugs

Sj-Si commented:
## Brief

This is an update of the current state of this PR.

## Changes Since Initial PR

- Changed layout of extra network tabs.
- Added a directory tree viewer in the left 1/4 of each extra network tab.
- Moved search/sort/refresh buttons inside of each tab.
  - Split the functionality of these items so that they are associated with each tab individually. So now your filters/sorting in the Lora tab do not apply to the Checkpoints tab (or any other tabs).
- Updated documentation.

## Known Issues

- Clicking the refresh button after adding a file to a directory will refresh the page but it won't show any files in the `Cards` pane until the user refreshes the window via the browser reload button or switches to another tab then back again. I don't really know why this isn't working. It is an odd one. It seems to only affect the tab in which the new file was created, too. So if I add a new file to the `Lora` directory, only the `Lora` tab will be bugged.

## Demo

At the bottom of this comment is a video showing the state of this PR as of the time of this comment.

Below is an explanation of what I am showing in the video:

1. Scroll bar in card pane.
2. Clickable directories in tree pane.
  - Clicking on a directory in tree pane will select it (it gets highlighted) and add the directory to the filter at the top of the tree pane. The filter applies to the card pane.
3. Scroll bar in tree pane.
4. Clicking on a file in the tree/card pane will add/remove it from the prompt.
5. Demonstrating the search bar at top of tree pane to search for files with `by_path` in their file/folder name.
6. Showing the new `Sort Mode` and `Sort Direction` buttons.
  - The video capture doesn't show the tooltips on hover. When I hover over these buttons, they are displaying their sort mode description ("Sort by path", "Sort by filename", "Sort by date created", and "Sort by date modified"). The Sort Direction button just switched between "Ascending" and "Descending".
  - The `by_dc` and `by_dm` directories are shown being sorted by Date Created and Date Modified.
7. Showing a super long filename being truncated with an ellipsis (`...`).
8. Showing how each tab's search/sort are separated.

One thing I missed in this video is that you can click on the chevron icons in the tree view to collapse/expand directories without selecting them to add to the filter.

https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/fbb46b0d-c6e3-44d6-881e-bd7d97c0f197

Sj-Si committed:
Fix embeddings add/remove to/from prompt on click bugs.

Sj-Si committed:
Fix missing important style.

Sj-Si committed:
Fix various bugs including refresh bug.

AUTOMATIC1111 commented:
The diff is too big to be reviewed properly... I'll probably have to #yolo merge it, which I always regret.

Anyway, in the current state, `extra_networks_card_order_field`/`extra_networks_card_order` settings are ignored, which they shouldn't be.

There is also an issue of this being always enabled. For people who do not have have a rich directory structure, this is just a clear degradation because it takes 1/4 of space without being particularly useful. There should be an option of not including the tree view.

When tree view is disabled, `tree-list-controls` elements should still be available - I suggest you put them where they were in that case, into the row with tabs (it's possible to still keep unique `tree-list-controls` per page like you implemented it - just place one into the location when a tab is selected).

Sj-Si committed:
Add toggle button for tree view. Use default settings for sortmode and direction.

Sj-Si committed:
Add option in settings to enable/disable tree view by default.

Sj-Si commented:
> The diff is too big to be reviewed properly... I'll probably have to #yolo merge it, which I always regret.

Just let me know if there is anything I can do to help make the review process easier for you.

> Anyway, in the current state, `extra_networks_card_order_field`/`extra_networks_card_order` settings are ignored, which they shouldn't be.

Good catch. I just added those back in. Completely missed those last time.

> There is also an issue of this being always enabled. For people who do not have have a rich directory structure, this is just a clear degradation because it takes 1/4 of space without being particularly useful. There should be an option of not including the tree view.

Added an option in settings (along with a fancy new button) for this. The option in settings enables/disables tree view by default on launch (the default value for this setting is `disabled`). The new button allows the user to easily toggle the tree view.

> When tree view is disabled, `tree-list-controls` elements should still be available - I suggest you put them where they were in that case, into the row with tabs (it's possible to still keep unique `tree-list-controls` per page like you implemented it - just place one into the location when a tab is selected) (I also have to add that I really like how it looks, neat ,small, and properly aligned).

I moved the controls outside of the tree view. I was also thinking it didn't really feel right to have them in that tree view div.

Tree View Disabled:
![image](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/f186358b-6bc6-4f66-89a0-745561d9c4e2)
Tree View Enabled:
![image](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/146112734/a6297d2c-9a73-41de-90be-5ae62db5cf75)

Sorry if it feels like I'm pushing back a lot on these controls but I just feel like having them inside each tab looks cleaner. It also made some of the code a bit easier to separate as well.

One thing that I am unsure of is whether the icons for the sort mode and sort direction make their function clear enough without having to hover over them to view the tooltip. Like it makes sense to me but I don't really know what the average user of this UI is like...

> Additionally:
> 
> If you feel like it, there is working horizontal resize system, which works, like in txt2img tabs you can make the left column smaller or bigger. You may be able to reuse that for your purposes with treeview, if you'd like.

At one point I did look into using reusing that component but it looked like I'd only be able to use it if I added it via python using the `ResizeHandleRow` class. I still haven't really figured out how to properly mix the custom HTML/CSS with these python components so I kinda gave up on that.

AUTOMATIC1111 commented:
I was thinking more about something like this - to not take an extra row.

![firefox_Y975U7Nq0x](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/20920490/caf263b6-ed00-44b9-97a7-cc826ef47ebd)

Sj-Si commented:
@AUTOMATIC1111
Honestly because of the way that I set up the tree view pane, I don't think I would be able to move those buttons up there while keeping as many of the features as they have. Like... I'm sure it's possible but it would take a massive amount of work for me to figure out since those html elements live outside of the tab divs.

Sj-Si committed:
Fix broken scrollbars

Sj-Si committed:
Remove unnecessary template and simplify tree list.

Dasor92 commented:
is there a way to roll back having folders? When you have more than 2000 loras the new tree view is unusable. Also, the "change preview" button doesn't update automatically.

tsjslgy commented:
This is a controversial feature, I like to train lora myself, so there are different directories to divide and I know very well the location and role of each folder, the previous version can click on the folder and go directly to the directory, but now you have to click on the tree network several times to enter, very troublesome. Of course, I can only represent individuals, so if you need to add this function, please also add a button that does not enable this function, thank you
